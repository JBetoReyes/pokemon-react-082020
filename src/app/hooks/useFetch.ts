import { useState } from 'react';

export interface IFetchState<T> {
  loading: boolean;
  data: T | null;
  error?: Error | string;
}

const defaultState = <T>(): IFetchState<T> => ({
  loading: false,
  data: null,
});

const useFetch = <T>(url: string): IFetchState<T> => {
  const [state, setState] = useState<IFetchState<T>>(defaultState<T>());
  fetch(url)
    .then((response) => response.json())
    .then((data: T) =>
      setState({
        loading: true,
        data,
      })
    )
    .catch((e) => {
      setState({
        loading: false,
        data: state.data,
        error: e,
      });
    });
  return state;
};

export default useFetch;
