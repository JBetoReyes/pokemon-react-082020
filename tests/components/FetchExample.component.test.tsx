import React from 'react';
import FetchExample from '@components/hooks/FetchExample.component';
import { shallow } from 'enzyme';
import useFetch from '@app/hooks/useFetch';

jest.mock('@app/hooks/useFetch');
describe('FetchExample', () => {
  (useFetch as jest.Mock).mockReturnValue({
    data: [],
    loading: true,
  });
  let wrapper = shallow(<FetchExample />);
  beforeAll(() => {
    const mockFetch = fetch as typeof fetchMock;
    mockFetch.resetMocks();
  });
  beforeEach(() => {
    wrapper = shallow(<FetchExample />);
  });
  test('Should match with the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
  test('Should test that the received quote is rendered', () => {
    (useFetch as jest.Mock).mockReturnValue({
      data: [
        {
          quote_id: 1,
          quote: 'test quote',
          author: 'test author',
        },
      ],
      loading: false,
    });
    wrapper = shallow(<FetchExample />);
    expect(wrapper.find('h2').text().trim()).toBe('test quote');
    expect(wrapper.find('h3').text().trim()).toBe('test author');
  });
});
