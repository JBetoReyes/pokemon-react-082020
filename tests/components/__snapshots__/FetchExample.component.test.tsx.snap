// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FetchExample Should match with the snapshot 1`] = `
<Fragment>
  loading...
  <button
    onClick={[Function]}
    type="button"
  >
    Next
  </button>
  <button
    onClick={[Function]}
    type="button"
  >
    Previous
  </button>
</Fragment>
`;
